### DISCUSSION TITLE
UART baudrate 100000

### DISCUSSION THREAD

**Discussion by swiftimundo** [micropython/micropython#12640] [RP2040 / Pico]
I'm trying to use an RC receiver's SBUS with Pi Pico. The receiver's SBUS has a baudrate of 100000, but I have yet to be able to receive data from it with Pi Pico. Is RP2040 able to communicate at 100000 baud at its default operating frequency?

**Comment by robert-hh** [comment]
Yes. I have tested UART at 4_000_000 baud for data loss. And it worked fine. Just take care that rxbuf is larger than the longest message which you expect.

**Reply by swiftimundo** [reply]
Good to know. My rxbuf=250, which I believe should be larger than my longest message.

**Reply by robert-hh** [reply]
rxbuf=256 is the default.

**Reply by swiftimundo** [reply]
ahh ok, would an low rxbuf prevent me from receiving anything? It seems like I may not be receiving any bytes.

**Reply by robert-hh** [reply]
No. You can set it to any value >= 32. You should receive data, if it's properly connected with GND and RX.

**Reply by swiftimundo** [reply]
yea I'm pretty sure it is... I think my next step is to get my scope out and see that it's actually transmitting and my logic level shifter is working.. 
Thanks

**Comment by GitHubsSilverBullet** [comment]
> I'm trying to use an RC receiver's SBUS with Pi Pico. The receiver's SBUS has a baudrate of 100000, but I have yet to be able to receive data from it with Pi Pico. Is RP2040 able to communicate at 100000 baud at its default operating frequency?

The SBUS protocol is not a valid protocol for a standard UART.
I'd be very surprised if you could receive any sensible data using a standard UART. Because that would require a startbit (low), 5…8 databits, maybe a parity bit and definitely a stop bit(high).
SBUS could send you up to 16 bits.

**Reply by swiftimundo** [reply]
ahh thats annoying. I've seen a couple other implementations on other boards done with UART, so I assumed it was possible.

**Reply by GitHubsSilverBullet** [reply]
Invert the signal. Electrically. Maybe that's configurable on the RP2040. 
Then use the (pretty unusual) 8-E-2 setting.
If the regular UART can't do that, then there's always the PIO approach.

**Reply by GitHubsSilverBullet** [reply]
Just checked the datasheet. You could try to do a specific GPIO configuration for the Rx/Tx Pins.
```python
IO_BANK0_BASE = const(0x40014000)
x = mem32[IO_BANK0_BASE + gpionum*8 + 4]
x = (x & ~(3<<16)) | (1<<16)  # INOVER: invert the peri input
x = (x & ~(3<<12)) | (1<<12)  # OEOVER: drive output enable from inverse of peripheral signal selected by funcsel
mem32[IO_BANK0_BASE + gpionum*8 + 4] = x
```
should invert the signals.
This is NOT tested, just from datasheet and hacked pretty dirty.

**Reply by swiftimundo** [reply]
Thanks for the info, sorry for not responding sooner. 
Here are my UART init lines
```python
sbus = UART(1, 100000)
sbus.init(bits=8, parity=0, stop=2, invert=UART.INV_TX | UART.INV_RX)
``` 

Since micropython now supports inverting on UART, is it necessary to configure the GPIO like you did? I should note I'm still not able to read any data, though I have confirmed with my scope that data is being passed.

**Reply by robert-hh** [reply]
You have to repeat the baudrate in the init() call, otherwise it will be set to the default of 115200. You can as well use all init arguments in the instantiation call.

**Reply by GitHubsSilverBullet** [reply]
> Since micropython now supports inverting on UART,


Ahh, didn't know that.  Something new to learn every day.
And no, that somewhat messy setup won't be necessary anymore.

**Reply by robert-hh** [reply]
I recalled that wrong for the rp2 port. If baudrate is not set, the previously set value rules.

**Reply by swiftimundo** [reply]
FINALLY got output! I had switched to UART1 because UART0 wasnt working for me... but I just switched back and got a flood of data! Now to decode. 
Thank you both for your help, not sure which to mark as the answer because I got a piece of the puzzle from each response. 

Here's what I ended up with. 
```python
sbus = UART(0, 100000, rx=17)
sbus.init(baudrate=100000, rx=17, bits=8, parity=0, stop=2, invert=UART.INV_RX)
```
